www/js/lib/popovers.js

```diff
@@ -0,0 +1,455 @@
+    const uriComponent = uiUtil.removeUrlParameters(href);
+    const zimURL = uiUtil.deriveZimUrlFromRelativeUrl(uriComponent, baseUrl);
+    console.debug('Previewing ' + zimURL);
+    // Do a binary search in the URL index to get the directory entry for the requested article
+    return archive.getDirEntryByPath(zimURL).then(function (dirEntry) {
+        const readArticle = function (dirEntry) {
+            // Wrap legacy callback-based code in a Promise
+            return new Promise((resolve, reject) => {
+                // As we're reading Wikipedia articles, we can assume that they are UTF-8 encoded HTML data
+                archive.readUtf8File(dirEntry, function (fileDirEntry, htmlArticle) {
+                    const parser = new DOMParser();
+                    const doc = parser.parseFromString(htmlArticle, 'text/html');
+                    const articleBody = doc.body;
+                    if (articleBody) {
+                        // Establish the popup balloon's base URL and the absolute path for calculating the ZIM URL of links and images
+                        const balloonBaseURL = encodeURI(fileDirEntry.namespace + '/' + fileDirEntry.url.replace(/[^/]+$/, ''));
+                        const docUrl = new URL(articleDocument.location.href);
+                        const rootRelativePathPrefix = docUrl.pathname.replace(/([^.]\.zim\w?\w?\/).+$/i, '$1');
+                        let balloonString = '';
+                        // Remove all standalone style elements, because their content is shown by both innerText and textContent
+                        const styleElements = Array.from(articleBody.querySelectorAll('style'));
+                        styleElements.forEach(style => {
+                            style.parentNode.removeChild(style);
+                        });
+                        const paragraphs = Array.from(articleBody.querySelectorAll('p'));
+                        // Filter out empty paragraphs or those with less than 50 characters
+                        const nonEmptyParagraphs = paragraphs.filter(para => {
+                            const text = para.innerText.trim();
+                            return !/^\s*$/.test(text) && text.length >= 50;
+                        });
+                        if (nonEmptyParagraphs.length > 0) {
+                            let cumulativeCharCount = 0;
+                            // Add enough paras to complete the word count
+                            for (let i = 0; i < nonEmptyParagraphs.length; i++) {
+                                // Get the character count: to fill the larger box we need ~850 characters (815 plus leeway)
+                                const plainText = nonEmptyParagraphs[i].innerText;
+                                cumulativeCharCount += plainText.length;
+                                // In ServiceWorker mode, we need to transform the URLs of any links in the paragraph
```

Comment:
Also a candidate for a helper function, `transformServiceWorkerLinks`.

I've found that nice readability 'hack' is to even define these helpers inline inside the main function, especially if you're worried that they won't be useful at the top level. Something like:

```
function bigComplicatedFunction(data) {
  function preProcessFile(file) {
    // Lots of code
  }

  function rewriteLinks(body, links) {
    // more code ...
   }

   // Main function code
   data.forEach(() => {
      var foo = Bar.thing()
      preProcessFile(foo);
    
     if (foo == 'baz') {
        rewriteLinks(foo, data);
     }
  }
}
```